commit 7bea5e5e5701de0b90b6c422c954337ce860bb0f
Merge: 182a1c9 c8a95ec
Author: Scott Lahteine <sourcetree@thinkyhead.com>
Date:   Wed Dec 7 03:21:07 2016 -0600

    README up-to-date

diff --git a/Marlin/Marlin_main.cpp b/Marlin/Marlin_main.cpp
index 8ca8a1f..baa8c2f 100755
--- a/Marlin/Marlin_main.cpp
+++ b/Marlin/Marlin_main.cpp
@@ -9972,7 +9972,7 @@ void manage_inactivity(bool ignore_stepper_queue/*=false*/) {
     // ---------------------------------------------------------
     static int homeDebounceCount = 0;   // poor man's debouncing count
     const int HOME_DEBOUNCE_DELAY = 2500;
-    if (!READ(HOME_PIN)) {
+    if (!IS_SD_PRINTING && !READ(HOME_PIN)) {
       if (!homeDebounceCount) {
         enqueue_and_echo_commands_P(PSTR("G28"));
         LCD_MESSAGEPGM(MSG_AUTO_HOME);

