commit e66ab97a3c4da32093926c1cdebb7d9ec47ba2e3
Merge: b4abb0a 244ea20
Author: Richard Wackerbarth <rkw@dataplex.net>
Date:   Mon Aug 3 07:15:32 2015 -0500

    Merge Bootscreen Text (PR#24)

diff --git a/Marlin/Marlin_main.cpp b/Marlin/Marlin_main.cpp
index 6f01871..0fdbf6e 100644
--- a/Marlin/Marlin_main.cpp
+++ b/Marlin/Marlin_main.cpp
@@ -666,7 +666,6 @@ void setup() {
   Config_RetrieveSettings();
 
   lcd_init();
-  _delay_ms(1000);  // wait 1sec to display the splash screen
 
   tp_init();    // Initialize temperature loop
   plan_init();  // Initialize planner;

