commit 3c1653207c2e9a73555147eb57a2317f5540e854
Merge: bea6739 ae00fad
Author: Richard Wackerbarth <rkw@dataplex.net>
Date:   Mon Jul 27 20:33:20 2015 -0500

    Merge Memory Watcher (PR#2398)

diff --git a/Marlin/Marlin_main.cpp b/Marlin/Marlin_main.cpp
index 09e559f..0597e2c 100644
--- a/Marlin/Marlin_main.cpp
+++ b/Marlin/Marlin_main.cpp
@@ -213,6 +213,7 @@
  * ************* SCARA End ***************
  *
  * ************ Custom codes - This can change to suit future G-code regulations
+ * M100 - Watch Free Memory (For Debugging Only)
  * M851 - Set probe's Z offset (mm above extruder -- The value will always be negative)
 
 
@@ -225,6 +226,10 @@
  *
  */
 
+#ifdef M100_FREE_MEMORY_WATCHER
+  void gcode_M100();
+#endif
+
 #ifdef SDSUPPORT
   CardReader card;
 #endif
@@ -414,6 +419,8 @@ void plan_arc(float target[NUM_AXIS], float *offset, uint8_t clockwise);
 
 bool setTargetedHotend(int code);
 
+void serial_echopair_P(const char *s_P, int v)           { serialprintPGM(s_P); SERIAL_ECHO(v); }
+void serial_echopair_P(const char *s_P, long v)          { serialprintPGM(s_P); SERIAL_ECHO(v); }
 void serial_echopair_P(const char *s_P, float v)         { serialprintPGM(s_P); SERIAL_ECHO(v); }
 void serial_echopair_P(const char *s_P, double v)        { serialprintPGM(s_P); SERIAL_ECHO(v); }
 void serial_echopair_P(const char *s_P, unsigned long v) { serialprintPGM(s_P); SERIAL_ECHO(v); }
@@ -5372,6 +5379,12 @@ void process_next_command() {
           break;
       #endif // ENABLE_AUTO_BED_LEVELING && Z_PROBE_REPEATABILITY_TEST
 
+      #ifdef M100_FREE_MEMORY_WATCHER
+        case 100:
+          gcode_M100();
+          break;
+      #endif
+
       case 104: // M104
         gcode_M104();
         break;

