commit 316469e76a1dd3763e01c6fb5a564d22ea09c8f6
Merge: 3674614 42465ee
Author: Richard Wackerbarth <rkw@dataplex.net>
Date:   Sun Jun 28 08:22:29 2015 -0500

    Clean up comments and spacing

diff --git a/Marlin/Marlin_main.cpp b/Marlin/Marlin_main.cpp
index 70fd2cf..192d6ea 100644
--- a/Marlin/Marlin_main.cpp
+++ b/Marlin/Marlin_main.cpp
@@ -3805,7 +3805,7 @@ inline void gcode_M81() {
 inline void gcode_M82() { axis_relative_modes[E_AXIS] = false; }
 
 /**
- * M82: Set E codes relative while in Absolute Coordinates (G90) mode
+ * M83: Set E codes relative while in Absolute Coordinates (G90) mode
  */
 inline void gcode_M83() { axis_relative_modes[E_AXIS] = true; }
 
@@ -6153,6 +6153,9 @@ void mesh_plan_buffer_line(float x, float y, float z, const float e, float feed_
 
 /**
  * Prepare a single move and get ready for the next one
+ *
+ * (This may call plan_buffer_line several times to put
+ *  smaller moves into the planner for DELTA or SCARA.)
  */
 void prepare_move() {
   clamp_to_software_endstops(destination);

