commit 0e074d40505b7124cf80da6c91c5845ecf901eed
Merge: 08b722b f90a947
Author: Scott Lahteine <thinkyhead@users.noreply.github.com>
Date:   Thu Mar 5 02:14:34 2015 -0800

    Merge pull request #1561 from OhmEye/Development
    
    Fix autoretract

diff --git a/Marlin/Marlin_main.cpp b/Marlin/Marlin_main.cpp
index b065b82..e968392 100644
--- a/Marlin/Marlin_main.cpp
+++ b/Marlin/Marlin_main.cpp
@@ -1428,10 +1428,10 @@ void process_commands()
             if(autoretract_enabled)
             if( !(code_seen('X') || code_seen('Y') || code_seen('Z')) && code_seen('E')) {
               float echange=destination[E_AXIS]-current_position[E_AXIS];
-              if((echange<-MIN_RETRACT && !retracted) || (echange>MIN_RETRACT && retracted)) { //move appears to be an attempt to retract or recover
+              if((echange<-MIN_RETRACT && !retracted[active_extruder]) || (echange>MIN_RETRACT && retracted[active_extruder])) { //move appears to be an attempt to retract or recover
                   current_position[E_AXIS] = destination[E_AXIS]; //hide the slicer-generated retract/recover from calculations
                   plan_set_e_position(current_position[E_AXIS]); //AND from the planner
-                  retract(!retracted);
+                  retract(!retracted[active_extruder]);
                   return;
               }
             }

