commit d4ba23bc284e023a82dafee34378ee4bfd7157db
Merge: 3b193d1 477b6fa
Author: nothinman <martin@lukasik.name>
Date:   Mon Feb 17 11:27:05 2014 +0000

    Merge pull request #770 from josefpavlik/Marlin_v1
    
    Marlin v1

diff --git a/Marlin/Marlin_main.cpp b/Marlin/Marlin_main.cpp
index 2b8138f..498adfd 100644
--- a/Marlin/Marlin_main.cpp
+++ b/Marlin/Marlin_main.cpp
@@ -1080,6 +1080,10 @@ static void homeaxis(int axis) {
   }
 }
 #define HOMEAXIS(LETTER) homeaxis(LETTER##_AXIS)
++void refresh_cmd_timeout(void)
+{
+  previous_millis_cmd = millis();
+}
 
 void process_commands()
 {

