commit 4823791eb0f553306bf4db469b729a45a2d5f7e5 (from 5e5d2508328e78fbddc3e02104dc3df448d2f196)
Merge: 5e5d250 89a8456
Author: Scott Lahteine <thinkyhead@users.noreply.github.com>
Date:   Fri Mar 25 00:40:51 2016 -0700

    Merge pull request #3228 from esenapaj/patch-1
    
    Fix spelling

diff --git a/Marlin/Marlin_main.cpp b/Marlin/Marlin_main.cpp
index fd49ab6..f9707da 100644
--- a/Marlin/Marlin_main.cpp
+++ b/Marlin/Marlin_main.cpp
@@ -3084,7 +3084,7 @@ inline void gcode_G28() {
           #if ENABLED(DELTA)
             // Avoid probing the corners (outside the round or hexagon print surface) on a delta printer.
             float distance_from_center = sqrt(xProbe * xProbe + yProbe * yProbe);
-            if (distance_from_center > DELTA_PROBABLE_RADIUS) continue;
+            if (distance_from_center > DELTA_PROBEABLE_RADIUS) continue;
           #endif //DELTA
 
           ProbeAction act;

