commit b7928a000a586fdbd39f12b909e2d0da6dde49cf (from 37b857e8ecb1bf50e59c2719d1d609986420dc61)
Merge: 5cba0b4 37b857e
Author: Scott Lahteine <thinkyhead@users.noreply.github.com>
Date:   Tue Mar 8 03:54:07 2016 -0800

    Merge pull request #3108 from AnHardt/M110+
    
    M110 and M29 rework

diff --git a/Marlin/Marlin_main.cpp b/Marlin/Marlin_main.cpp
index bd4c648..fb1d3b3 100644
--- a/Marlin/Marlin_main.cpp
+++ b/Marlin/Marlin_main.cpp
@@ -871,7 +871,7 @@ void get_command() {
           }
           // if no errors, continue parsing
         }
-        else if (npos == command) {
+        else {
           gcode_line_error(PSTR(MSG_ERR_NO_CHECKSUM));
           return;
         }

