commit 588ed70d4a9618807390269d943b5e58aa001933 (from 075386e7febf5d88d85c8421c06f10bcfe7c80b1)
Merge: 7fb8d9f 075386e
Author: Scott Lahteine <sourcetree@thinkyhead.com>
Date:   Thu Apr 23 21:18:48 2015 -0700

    Latest upstream commits, merged

diff --git a/Marlin/Marlin_main.cpp b/Marlin/Marlin_main.cpp
index a06bf78..c45dc76 100644
--- a/Marlin/Marlin_main.cpp
+++ b/Marlin/Marlin_main.cpp
@@ -1030,6 +1030,10 @@ static void axis_is_at_home(int axis) {
     current_position[axis] = base_home_pos(axis) + home_offset[axis];
     min_pos[axis] = base_min_pos(axis) + home_offset[axis];
     max_pos[axis] = base_max_pos(axis) + home_offset[axis];
+
+    #if defined(ENABLE_AUTO_BED_LEVELING) && Z_HOME_DIR < 0
+      if (axis == Z_AXIS) current_position[Z_AXIS] += zprobe_zoffset;
+    #endif
   }
 }
 
@@ -2037,10 +2041,6 @@ inline void gcode_G28() {
     if (code_seen(axis_codes[Z_AXIS]) && code_has_value())
       current_position[Z_AXIS] = code_value();
 
-    #if defined(ENABLE_AUTO_BED_LEVELING) && (Z_HOME_DIR < 0)
-      if (home_all_axis || homeZ) current_position[Z_AXIS] += zprobe_zoffset;  // Add Z_Probe offset (the distance is negative)
-    #endif
-
     sync_plan_position();
 
   #endif // else DELTA

