commit fdb4ddbdea093d468492cf73d576d9d6663da3be (from c89b8368ee678b693362b0833b1f425caab4af53)
Merge: c89b836 0b57a2a
Author: Chris Roadfeldt <chris@roadfeldt.com>
Date:   Wed Apr 1 00:52:08 2015 -0500

    Merge remote-tracking branch 'MarlinFirmware/Development' into Development

diff --git a/Marlin/Marlin_main.cpp b/Marlin/Marlin_main.cpp
index ce694e2..d95ad89 100644
--- a/Marlin/Marlin_main.cpp
+++ b/Marlin/Marlin_main.cpp
@@ -843,7 +843,7 @@ void get_command()
         int hours, minutes;
         minutes=(t/60)%60;
         hours=t/60/60;
-        sprintf_P(time, PSTR("%i hours %i minutes"),hours, minutes);
+        sprintf_P(time, PSTR("%i "MSG_END_HOUR" %i "MSG_END_MINUTE),hours, minutes);
         SERIAL_ECHO_START;
         SERIAL_ECHOLN(time);
         lcd_setstatus(time, true);

