commit 956fb2d2789719443a6c2402e43040baf1c4a6ce (from cfc6a3a87a5274dd38165026f24ee245f2bdc030)
Merge: cfc6a3a 624f23f
Author: Mehmet Sutas <msutas@gmail.com>
Date:   Mon Mar 9 11:05:49 2015 +0200

    Merge pull request #1 from MarlinFirmware/Development
    
    update

diff --git a/Marlin/Marlin_main.cpp b/Marlin/Marlin_main.cpp
index 9ce7611..77d1e82 100644
--- a/Marlin/Marlin_main.cpp
+++ b/Marlin/Marlin_main.cpp
@@ -5100,18 +5100,18 @@ void controllerFan()
   if ((millis() - lastMotorCheck) >= 2500) //Not a time critical function, so we only check every 2500ms
   {
     lastMotorCheck = millis();
-
-    if(!READ(X_ENABLE_PIN) || !READ(Y_ENABLE_PIN) || !READ(Z_ENABLE_PIN) || (soft_pwm_bed > 0)
+	
+    if((READ(X_ENABLE_PIN) == (X_ENABLE_ON)) || (READ(Y_ENABLE_PIN) == (Y_ENABLE_ON)) || (READ(Z_ENABLE_PIN) == (Z_ENABLE_ON)) || (soft_pwm_bed > 0)
     #if EXTRUDERS > 2
-       || !READ(E2_ENABLE_PIN)
+       || (READ(E2_ENABLE_PIN) == (E_ENABLE_ON))
     #endif
     #if EXTRUDER > 1
       #if defined(X2_ENABLE_PIN) && X2_ENABLE_PIN > -1
-       || !READ(X2_ENABLE_PIN)
+       || (READ(X2_ENABLE_PIN) == (X_ENABLE_ON))
       #endif
-       || !READ(E1_ENABLE_PIN)
+       || (READ(E1_ENABLE_PIN) == (E_ENABLE_ON))
     #endif
-       || !READ(E0_ENABLE_PIN)) //If any of the drivers are enabled...
+       || (READ(E0_ENABLE_PIN) == (E_ENABLE_ON))) //If any of the drivers are enabled...
     {
       lastMotor = millis(); //... set time to NOW so the fan will turn on
     }

