commit 64139f5e5a4129c5492f947101744593b0055dc7 (from 7562183efc719b25482b7e82bee56c1a3bcb8bbf)
Merge: b81021f 7562183
Author: Neil Darlow <neil@darlow.co.uk>
Date:   Thu Dec 4 21:09:03 2014 +0000

    Merge https://github.com/ErikZalm/Marlin into Marlin_v1

diff --git a/Marlin/Marlin_main.cpp b/Marlin/Marlin_main.cpp
index 616b7a1..e8d800e 100644
--- a/Marlin/Marlin_main.cpp
+++ b/Marlin/Marlin_main.cpp
@@ -1896,6 +1896,8 @@ void process_commands()
         }
         lcd_ignore_click(false);
       }else{
+          if (!lcd_detected())
+            break;
         while(!lcd_clicked()){
           manage_heater();
           manage_inactivity();

