commit b81021f4759a1e3244cef84e30dbb264b6bef936 (from 058e4465313191bf769e00a347ceccb4fbd1112c)
Merge: 6c5a460 058e446
Author: Neil Darlow <neil@darlow.co.uk>
Date:   Tue Oct 21 12:06:56 2014 +0100

    Merge https://github.com/ErikZalm/Marlin into Marlin_v1

diff --git a/Marlin/Marlin_main.cpp b/Marlin/Marlin_main.cpp
index 94b3d26..3bd8e31 100644
--- a/Marlin/Marlin_main.cpp
+++ b/Marlin/Marlin_main.cpp
@@ -1878,6 +1878,8 @@ void process_commands()
           lcd_update();
         }
       }else{
+          if (!lcd_detected())
+            break;
         while(!lcd_clicked()){
           manage_heater();
           manage_inactivity();

