commit c3cdaceab3e5d5555f673773cdcbedc47f6b5e6e (from 5d5909fc2532d9ebb3bbe0f54341ea5046baff5c)
Merge: 372e12f 5d5909f
Author: Robert F-C <rfairliecuninghame@yahoo.com>
Date:   Mon Apr 29 22:13:55 2013 +1000

    Merge remote-tracking branch 'upstream/Marlin_v1' into Marlin_v1

diff --git a/Marlin/Marlin_main.cpp b/Marlin/Marlin_main.cpp
index 04c3794..69f464b 100644
--- a/Marlin/Marlin_main.cpp
+++ b/Marlin/Marlin_main.cpp
@@ -1003,6 +1003,10 @@ void process_commands()
             break;
           }
         }
+      #if FAN_PIN > -1
+        if (pin_number == FAN_PIN)
+          fanSpeed = pin_status;
+      #endif
         if (pin_number > -1)
         {
           pinMode(pin_number, OUTPUT);

