commit 06906bc1c8e520683d88e27f8ad376012167d8c0
Merge: b93cddf 318318b
Author: Richard Wackerbarth <rkw@dataplex.net>
Date:   Mon Jun 15 20:18:21 2015 -0500

    Merge (PR#2295)

diff --git a/Marlin/pins.h b/Marlin/pins.h
index 297dc86..4893cf5 100644
--- a/Marlin/pins.h
+++ b/Marlin/pins.h
@@ -34,7 +34,9 @@
   #include "pins_SETHI.h"
 #elif MB(RAMPS_OLD)
   #include "pins_RAMPS_OLD.h"
-#elif MB(RAMPS_13_EFB) || MB(RAMPS_13_EEB) || MB(RAMPS_13_EFF) || MB(RAMPS_13_EEF) || MB(RAMPS_13_SF)
+#elif MB(RAMPS_13_EFB)
+  #include "pins_RAMPS_13_EFB.h"
+#elif MB(RAMPS_13_EEB) || MB(RAMPS_13_EFF) || MB(RAMPS_13_EEF) || MB(RAMPS_13_SF)
   #include "pins_RAMPS_13.h"
 #elif MB(GEN6)
   #include "pins_GEN6.h"
diff --git a/Marlin/pins_AZTEEG_X3.h b/Marlin/pins_AZTEEG_X3.h
index e961e7a..9da88bc 100644
--- a/Marlin/pins_AZTEEG_X3.h
+++ b/Marlin/pins_AZTEEG_X3.h
@@ -2,9 +2,7 @@
  * AZTEEG_X3 Arduino Mega with RAMPS v1.3 pin assignments
  */
 
-#define IS_RAMPS_EFB
-
-#include "pins_RAMPS_13.h"
+#include "pins_RAMPS_13_EFB.h"
 
 //LCD Pins//
 
diff --git a/Marlin/pins_BAM_DICE_DUE.h b/Marlin/pins_BAM_DICE_DUE.h
index 2ec354e..29d6109 100644
--- a/Marlin/pins_BAM_DICE_DUE.h
+++ b/Marlin/pins_BAM_DICE_DUE.h
@@ -2,9 +2,7 @@
  * BAM&DICE Due (Arduino Mega) pin assignments
  */
 
-#define IS_RAMPS_EFB
-
-#include "pins_RAMPS_13.h"
+#include "pins_RAMPS_13_EFB.h"
 
 #undef TEMP_0_PIN
 #undef TEMP_1_PIN
diff --git a/Marlin/pins_FELIX2.h b/Marlin/pins_FELIX2.h
index 2fa40e2..46c414f 100644
--- a/Marlin/pins_FELIX2.h
+++ b/Marlin/pins_FELIX2.h
@@ -2,9 +2,7 @@
  * FELIXprinters v2.0/3.0 (RAMPS v1.3) pin assignments
  */
 
-#define IS_RAMPS_EFB
-
-#include "pins_RAMPS_13.h"
+#include "pins_RAMPS_13_EFB.h"
 
 #undef HEATER_1_PIN
 #define HEATER_1_PIN        7 // EXTRUDER 2
diff --git a/Marlin/pins_MKS_BASE.h b/Marlin/pins_MKS_BASE.h
index fd4d2f1..3cf5be1 100644
--- a/Marlin/pins_MKS_BASE.h
+++ b/Marlin/pins_MKS_BASE.h
@@ -2,9 +2,7 @@
  * MKS BASE 1.0 – Arduino Mega2560 with RAMPS v1.4 pin assignments
  */
 
-#define IS_RAMPS_EFB
-
-#include "pins_RAMPS_13.h"
+#include "pins_RAMPS_13_EFB.h"
 
 #undef HEATER_1_PIN
 #define HEATER_1_PIN        7
diff --git a/Marlin/pins_RAMPS_13.h b/Marlin/pins_RAMPS_13.h
index a99be9c..9f4313f 100644
--- a/Marlin/pins_RAMPS_13.h
+++ b/Marlin/pins_RAMPS_13.h
@@ -77,7 +77,7 @@
   #define FILRUNOUT_PIN        4
 #endif
 
-#if MB(RAMPS_13_EFB) || MB(RAMPS_13_EFF) || defined(IS_RAMPS_EFB)
+#if MB(RAMPS_13_EFF) || defined(IS_RAMPS_EFB)
   #define FAN_PIN            9 // (Sprinter config)
   #if MB(RAMPS_13_EFF)
     #define CONTROLLERFAN_PIN  -1 // Pin used for the fan to cool controller
@@ -102,7 +102,7 @@
   #define HEATER_0_PIN       10   // EXTRUDER 1
 #endif
 
-#if MB(RAMPS_13_EFB) || MB(RAMPS_13_SF) || defined(IS_RAMPS_EFB)
+#if MB(RAMPS_13_SF) || defined(IS_RAMPS_EFB)
   #define HEATER_1_PIN       -1
 #else
   #define HEATER_1_PIN       9    // EXTRUDER 2 (FAN On Sprinter)
diff --git a/Marlin/pins_RAMPS_13_EFB.h b/Marlin/pins_RAMPS_13_EFB.h
new file mode 100644
index 0000000..c75acd8
--- /dev/null
+++ b/Marlin/pins_RAMPS_13_EFB.h
@@ -0,0 +1,9 @@
+/**
+ * Arduino Mega with RAMPS v1.3 pin assignments
+ *
+ *  RAMPS_13_EFB (Extruder, Fan, Bed)
+ */
+
+#define IS_RAMPS_EFB
+
+#include "pins_RAMPS_13.h"
\ No newline at end of file
