commit 096046c09fad3584b170ba64ba021e546fbd584e (from 960de3411b66b4c748934260b11f024a0138e35a)
Merge: 960de34 1183c10
Author: Richard Wackerbarth <rkw@dataplex.net>
Date:   Sun Jul 26 20:32:10 2015 -0500

    Merge Improve Makefile use of avrdude (PR#2477)

diff --git a/Marlin/Makefile b/Marlin/Makefile
index b22cca6..684e949 100644
--- a/Marlin/Makefile
+++ b/Marlin/Makefile
@@ -350,11 +350,11 @@ LDFLAGS = -lm
 AVRDUDE_PORT = $(UPLOAD_PORT)
 AVRDUDE_WRITE_FLASH = -Uflash:w:$(BUILD_DIR)/$(TARGET).hex:i
 ifeq ($(shell uname -s), Linux)
-AVRDUDE_CONF = $(ARDUINO_INSTALL_DIR)/hardware/tools/avrdude.conf
+AVRDUDE_CONF = /etc/avrdude/avrdude.conf
 else
 AVRDUDE_CONF = $(ARDUINO_INSTALL_DIR)/hardware/tools/avr/etc/avrdude.conf
 endif
-AVRDUDE_FLAGS = -q -q -D -C$(AVRDUDE_CONF) \
+AVRDUDE_FLAGS = -D -C$(AVRDUDE_CONF) \
 	-p$(MCU) -P$(AVRDUDE_PORT) -c$(AVRDUDE_PROGRAMMER) \
 	-b$(UPLOAD_RATE)
 

