commit 122bdd187f9df02a6d9b59495630e40da307e82c
Merge: 9045ea3 f394c06
Author: Scott Lahteine <sourcetree@thinkyhead.com>
Date:   Fri Mar 25 01:15:38 2016 -0700

    Merge branch 'RCBugFix' into RC

diff --git a/Marlin/ultralcd.cpp b/Marlin/ultralcd.cpp
index 108ce4b..2dd247a 100644
--- a/Marlin/ultralcd.cpp
+++ b/Marlin/ultralcd.cpp
@@ -1114,7 +1114,7 @@ static void lcd_control_menu() {
     enqueue_and_echo_command_now(cmd);
   }
 
-#endif PIDTEMP || PIDTEMPBED
+#endif //PIDTEMP || PIDTEMPBED
 
 #if ENABLED(PIDTEMP)
 
