commit 2bb5cf5291208f72053fe880bab2cb18534f9c01
Merge: 6d98f47 59cfc84
Author: Scott Lahteine <thinkyhead@users.noreply.github.com>
Date:   Tue Apr 26 17:14:36 2016 -0700

    Merge pull request #3622 from floyd871/patch-1

diff --git a/Marlin/language_de.h b/Marlin/language_de.h
index 06d718c..adccd1f 100644
--- a/Marlin/language_de.h
+++ b/Marlin/language_de.h
@@ -103,7 +103,7 @@
 #define MSG_VTRAV_MIN                       "VTrav min"
 #define MSG_AMAX                            "A max " // space by purpose
 #define MSG_A_RETRACT                       "A Retract"
-#define MSG_A_TRAVEL                        "A Rueckzug"
+#define MSG_A_TRAVEL                        "A Rückzug"
 #define MSG_XSTEPS                          "X steps/mm"
 #define MSG_YSTEPS                          "Y steps/mm"
 #define MSG_ZSTEPS                          "Z steps/mm"
@@ -131,7 +131,7 @@
 #define MSG_RESUMING                        "Druck geht weiter"
 #define MSG_PRINT_ABORTED                   "Druck abgebrochen"
 #define MSG_NO_MOVE                         "Motoren Eingesch."
-#define MSG_KILLED                          "KILLED."
+#define MSG_KILLED                          "BEENDET."
 #define MSG_STOPPED                         "ANGEHALTEN."
 #define MSG_CONTROL_RETRACT                 "Retract mm"
 #define MSG_CONTROL_RETRACT_SWAP            "Wechs. Retract mm"
@@ -154,9 +154,9 @@
 #define MSG_HEATING_FAILED_LCD              "Heizen fehlgesch."
 #define MSG_ERR_REDUNDANT_TEMP              "Redund. Temperaturabw."
 #define MSG_THERMAL_RUNAWAY                 "Temp. n. erreicht"
-#define MSG_ERR_MAXTEMP                     "Temp. ueberschritten"
+#define MSG_ERR_MAXTEMP                     "Temp. überschritten"
 #define MSG_ERR_MINTEMP                     "Temp. unterschritten"
-#define MSG_ERR_MAXTEMP_BED                 "Temp. Bett ueberschr."
+#define MSG_ERR_MAXTEMP_BED                 "Temp. Bett überschr."
 #define MSG_ERR_MINTEMP_BED                 "Temp. Bett unterschr."
 #define MSG_END_HOUR                        "Stunden"
 #define MSG_END_MINUTE                      "Minuten"
