commit 3eea5222cf8ae99fbe8dd3ddca5957dca57a1d76
Merge: 10733a2 6d0e913
Author: Richard Wackerbarth <rkw@dataplex.net>
Date:   Sat Jun 6 19:00:27 2015 -0500

    Merge PR #1767

diff --git a/Marlin/boards.h b/Marlin/boards.h
index f854b8f..1748ac4 100644
--- a/Marlin/boards.h
+++ b/Marlin/boards.h
@@ -42,7 +42,7 @@
 #define BOARD_GEN3_MONOLITHIC   22   // Gen3 Monolithic Electronics
 #define BOARD_MEGATRONICS       70   // Megatronics
 #define BOARD_MEGATRONICS_2     701  // Megatronics v2.0
-#define BOARD_MEGATRONICS_1     702  // Minitronics v1.0
+#define BOARD_MINITRONICS       702  // Minitronics v1.0/1.1
 #define BOARD_MEGATRONICS_3     703  // Megatronics v3.0
 #define BOARD_OMCA_A            90   // Alpha OMCA board
 #define BOARD_OMCA              91   // Final OMCA board
diff --git a/Marlin/pins.h b/Marlin/pins.h
index d5ce7ee..9da667c 100644
--- a/Marlin/pins.h
+++ b/Marlin/pins.h
@@ -86,8 +86,8 @@
   #include "pins_GEN3_MONOLITHIC.h"
 #elif MB(MEGATRONICS)
   #include "pins_MEGATRONICS.h"
-#elif MB(MEGATRONICS_1)
-  #include "pins_MEGATRONICS_1.h"
+#elif MB(MINITRONICS)
+  #include "pins_MINITRONICS.h"
 #elif MB(MEGATRONICS_2)
   #include "pins_MEGATRONICS_2.h"
 #elif MB(MEGATRONICS_3)
diff --git a/Marlin/pins_MEGATRONICS_1.h b/Marlin/pins_MINITRONICS.h
similarity index 97%
rename from Marlin/pins_MEGATRONICS_1.h
rename to Marlin/pins_MINITRONICS.h
index 1592dfe..eead2d4 100644
--- a/Marlin/pins_MEGATRONICS_1.h
+++ b/Marlin/pins_MINITRONICS.h
@@ -1,5 +1,5 @@
 /**
- * Minitronics v1.0 pin assignments
+ * Minitronics v1.0/1.1 pin assignments
  */
 
 #ifndef __AVR_ATmega1281__
