commit 78b7c486fdf83d14c148e0f0e4ddc494b7fe5c3f
Merge: e4c01f1 87a97f3
Author: Richard Wackerbarth <rkw@dataplex.net>
Date:   Sun Jul 26 20:42:11 2015 -0500

    Merge Translate additional Displays (PR#2529)

diff --git a/Marlin/language.h b/Marlin/language.h
index 1405252..ad04474 100644
--- a/Marlin/language.h
+++ b/Marlin/language.h
@@ -123,10 +123,6 @@
 #define MSG_INVALID_EXTRUDER                "Invalid extruder"
 #define MSG_INVALID_SOLENOID                "Invalid solenoid"
 #define MSG_ERR_NO_THERMISTORS              "No thermistors - no temperature"
-#define MSG_HEATING                         "Heating..."
-#define MSG_HEATING_COMPLETE                "Heating done."
-#define MSG_BED_HEATING                     "Bed Heating."
-#define MSG_BED_DONE                        "Bed done."
 #define MSG_M115_REPORT                     "FIRMWARE_NAME:Marlin " BUILD_VERSION " SOURCE_CODE_URL:" SOURCE_CODE_URL " PROTOCOL_VERSION:" PROTOCOL_VERSION " MACHINE_TYPE:" MACHINE_NAME " EXTRUDER_COUNT:" STRINGIFY(EXTRUDERS) " UUID:" MACHINE_UUID "\n"
 #define MSG_COUNT_X                         " Count X: "
 #define MSG_ERR_KILLED                      "Printer halted. kill() called!"
diff --git a/Marlin/language_de.h b/Marlin/language_de.h
index 57db43f..d8bdbf4 100644
--- a/Marlin/language_de.h
+++ b/Marlin/language_de.h
@@ -124,6 +124,10 @@
 #define MSG_ENDSTOP_ABORT                   "Endstop Abbr. Ein"
 #define MSG_END_HOUR                        "Stunden"
 #define MSG_END_MINUTE                      "Minuten"
+#define MSG_HEATING                         "Aufheizen..."
+#define MSG_HEATING_COMPLETE                "Aufgeheizt"
+#define MSG_BED_HEATING                     "Bett aufheizen"
+#define MSG_BED_DONE                        "Bett aufgeheizt"
 
 #ifdef DELTA_CALIBRATION_MENU
     #define MSG_DELTA_CALIBRATE             "Delta kalibrieren"
diff --git a/Marlin/language_en.h b/Marlin/language_en.h
index f2ca91b..82df359 100644
--- a/Marlin/language_en.h
+++ b/Marlin/language_en.h
@@ -420,6 +420,18 @@
 #ifndef MSG_END_MINUTE
 #define MSG_END_MINUTE                      "minutes"
 #endif
+#ifndef MSG_HEATING
+#define MSG_HEATING                         "Heating..."
+#endif
+#ifndef MSG_HEATING_COMPLETE
+#define MSG_HEATING_COMPLETE                "Heating done."
+#endif
+#ifndef MSG_BED_HEATING
+#define MSG_BED_HEATING                     "Bed Heating."
+#endif
+#ifndef MSG_BED_DONE
+#define MSG_BED_DONE                        "Bed done."
+#endif
 
 #ifdef DELTA_CALIBRATION_MENU
   #ifndef MSG_DELTA_CALIBRATE
