commit 88506a80c318f3b2a7b11bbca3751341f4e29580
Merge: b514d3c c00383f
Author: ErikZalm <erik@vdzalm.eu>
Date:   Mon Feb 24 22:26:31 2014 +0100

    Merge pull request #789 from smiler/Marlin_v1
    
    Fixed typo.

diff --git a/Marlin/temperature.cpp b/Marlin/temperature.cpp
index 66a097d..060d11f 100644
--- a/Marlin/temperature.cpp
+++ b/Marlin/temperature.cpp
@@ -250,7 +250,7 @@ void PID_autotune(float temp, int extruder, int ncycles)
               Kp = 0.6*Ku;
               Ki = 2*Kp/Tu;
               Kd = Kp*Tu/8;
-              SERIAL_PROTOCOLLNPGM(" Clasic PID ");
+              SERIAL_PROTOCOLLNPGM(" Classic PID ");
               SERIAL_PROTOCOLPGM(" Kp: "); SERIAL_PROTOCOLLN(Kp);
               SERIAL_PROTOCOLPGM(" Ki: "); SERIAL_PROTOCOLLN(Ki);
               SERIAL_PROTOCOLPGM(" Kd: "); SERIAL_PROTOCOLLN(Kd);
