commit 9045ea3adf0dc66008e5c73cb684e3654bf39665
Merge: 6f9442a f52f388
Author: Scott Lahteine <sourcetree@thinkyhead.com>
Date:   Fri Mar 25 01:12:37 2016 -0700

    Merge branch 'RCBugFix' into RC

diff --git a/Marlin/dogm_lcd_implementation.h b/Marlin/dogm_lcd_implementation.h
index 107b860..146dcd8 100644
--- a/Marlin/dogm_lcd_implementation.h
+++ b/Marlin/dogm_lcd_implementation.h
@@ -304,8 +304,10 @@ static void _draw_heater_status(int x, int heater) {
 static void lcd_implementation_status_screen() {
   u8g.setColorIndex(1); // black on white
 
-  // Symbols menu graphics, animated fan
-  u8g.drawBitmapP(9,1,STATUS_SCREENBYTEWIDTH,STATUS_SCREENHEIGHT, (blink % 2) && fanSpeeds[0] ? status_screen0_bmp : status_screen1_bmp);
+  #if HAS_FAN0
+    // Symbols menu graphics, animated fan
+    u8g.drawBitmapP(9, 1, STATUS_SCREENBYTEWIDTH, STATUS_SCREENHEIGHT, (blink % 2) && fanSpeeds[0] ? status_screen0_bmp : status_screen1_bmp);
+  #endif
 
   #if ENABLED(SDSUPPORT)
     // SD Card Symbol
