commit 960de3411b66b4c748934260b11f024a0138e35a
Merge: fd0e81b bcb7d5a
Author: Richard Wackerbarth <rkw@dataplex.net>
Date:   Sat Jul 25 21:18:32 2015 -0500

    Merge Temp Sanity Checks (PR#2514)

diff --git a/Marlin/SanityCheck.h b/Marlin/SanityCheck.h
index 6c4c77a..6c7858b 100644
--- a/Marlin/SanityCheck.h
+++ b/Marlin/SanityCheck.h
@@ -51,7 +51,7 @@
   #endif
 
   /**
-   * Options only for EXTRUDERS == 1
+   * Options only for EXTRUDERS > 1
    */
   #if EXTRUDERS > 1
 
@@ -281,24 +281,38 @@
   #endif
 
   /**
-   * Test required HEATER defines
+   * Test required HEATER and TEMP_SENSOR defines
    */
   #if EXTRUDERS > 3
     #if !HAS_HEATER_3
       #error HEATER_3_PIN not defined for this board.
+    #elif TEMP_SENSOR_3 == 0
+      #error TEMP_SENSOR_3 is required with 4 EXTRUDERS.
     #endif
   #elif EXTRUDERS > 2
     #if !HAS_HEATER_2
       #error HEATER_2_PIN not defined for this board.
+    #elif TEMP_SENSOR_2 == 0
+      #error TEMP_SENSOR_2 is required with 3 or more EXTRUDERS.
     #endif
   #elif EXTRUDERS > 1 || defined(HEATERS_PARALLEL)
     #if !HAS_HEATER_1
       #error HEATER_1_PIN not defined for this board.
     #endif
   #endif
+  #if TEMP_SENSOR_1 == 0
+    #if EXTRUDERS > 1
+      #error TEMP_SENSOR_1 is required with 2 or more EXTRUDERS.
+    #elif ENABLED(TEMP_SENSOR_1_AS_REDUNDANT)
+      #error TEMP_SENSOR_1 is required with TEMP_SENSOR_1_AS_REDUNDANT.
+    #endif
+  #endif
   #if !HAS_HEATER_0
     #error HEATER_0_PIN not defined for this board.
   #endif
+  #if TEMP_SENSOR_0 == 0
+    #error TEMP_SENSOR_0 is required.
+  #endif
 
   /**
    * Warnings for old configurations
