commit 9e603920c2883af00c918031fc00147557e71a48 (from b84db7ff5b71f593b85f2b94e90fbc5d203f5416)
Merge: b84db7f 59994bd
Author: Chris Roadfeldt <chris@roadfeldt.com>
Date:   Wed Apr 1 19:31:14 2015 -0500

    Merge branch 'Merge_cleanup' into Development

diff --git a/Marlin/Marlin_main.cpp b/Marlin/Marlin_main.cpp
index 5bc239d..456cfb0 100644
--- a/Marlin/Marlin_main.cpp
+++ b/Marlin/Marlin_main.cpp
@@ -2811,7 +2811,7 @@ inline void gcode_M42() {
       #error "You must have a Z_PROBE_PIN defined in order to enable calculation of Z-Probe repeatability."
     #endif
   #else
-    #if (Z_MIN_PIN == -1) &&
+    #if (Z_MIN_PIN == -1)
       #error "You must have a Z_MIN_PIN defined in order to enable calculation of Z-Probe repeatability."
     #endif
   #endif
diff --git a/Marlin/SanityCheck.h b/Marlin/SanityCheck.h
index 6c0c9fd..8ea9b2e 100644
--- a/Marlin/SanityCheck.h
+++ b/Marlin/SanityCheck.h
@@ -93,7 +93,7 @@
      * Require a Z Min pin
      */
     #if Z_MIN_PIN == -1
-      #if Z_PROBE_PIN == -1
+      #if Z_PROBE_PIN == -1 || (! defined (Z_PROBE_ENDSTOP) || defined (DISABLE_Z_PROBE_ENDSTOP)) // It's possible for someone to set a pin for the Z Probe, but not enable it.
         #ifdef Z_PROBE_REPEATABILITY_TEST
           #error You must have a Z_MIN or Z_PROBE endstop to enable Z_PROBE_REPEATABILITY_TEST.
         #else
diff --git a/Marlin/language.h b/Marlin/language.h
index f4a2d26..4a4698c 100644
--- a/Marlin/language.h
+++ b/Marlin/language.h
@@ -138,7 +138,7 @@
 #define MSG_Z_MIN                           "z_min: "
 #define MSG_Z_MAX                           "z_max: "
 #define MSG_Z2_MAX                          "z2_max: "
-#define MSG_Z_PROBE							"z_probe: "
+#define MSG_Z_PROBE                         "z_probe: "
 #define MSG_M119_REPORT                     "Reporting endstop status"
 #define MSG_ENDSTOP_HIT                     "TRIGGERED"
 #define MSG_ENDSTOP_OPEN                    "open"

