commit a0120a533cfc06178f2df0ab7c0b4d4e207bda41 (from 469843427deb25f07e83808b59a213709761e83c)
Merge: 4698434 4e14e15
Author: Richard Wackerbarth <rkw@dataplex.net>
Date:   Thu Jul 23 19:53:33 2015 -0500

    Merge CORE_XZ (PR#2503)

diff --git a/.travis.yml b/.travis.yml
index b05431c..9543fda 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -140,6 +140,16 @@ script:
   - sed -i 's/\/\/#define FILAMENT_LCD_DISPLAY/#define FILAMENT_LCD_DISPLAY/g' Marlin/Configuration.h
   - rm -rf .build/
   - DISPLAY=:1.0 ~/bin/arduino --verify --board marlin:avr:mega  Marlin/Marlin.ino
+  # Enable COREXY
+  - cp Marlin/Configuration.h.backup Marlin/Configuration.h
+  - sed -i 's/\/\/#define COREXY/#define COREXY/g' Marlin/Configuration.h
+  - rm -rf .build/
+  - DISPLAY=:1.0 ~/bin/arduino --verify --board marlin:avr:mega  Marlin/Marlin.ino
+  # Enable COREXZ
+  - cp Marlin/Configuration.h.backup Marlin/Configuration.h
+  - sed -i 's/\/\/#define COREXZ/#define COREXZ/g' Marlin/Configuration.h
+  - rm -rf .build/
+  - DISPLAY=:1.0 ~/bin/arduino --verify --board marlin:avr:mega  Marlin/Marlin.ino
   ######## Example Configurations ##############
   # Delta Config (generic)
   - cp Marlin/example_configurations/delta/generic/Configuration* Marlin/
diff --git a/Marlin/Configuration.h b/Marlin/Configuration.h
index 400d226..7d06e21 100644
--- a/Marlin/Configuration.h
+++ b/Marlin/Configuration.h
@@ -305,10 +305,10 @@ Here are some standard links for getting your machine calibrated:
 // @section machine
 
 // Uncomment this option to enable CoreXY kinematics
-// #define COREXY
+//#define COREXY
 
 // Uncomment this option to enable CoreXZ kinematics
-// #define COREXZ
+//#define COREXZ
 
 // Enable this option for Toshiba steppers
 //#define CONFIG_STEPPERS_TOSHIBA
diff --git a/Marlin/stepper.cpp b/Marlin/stepper.cpp
index 14f0475..f74f2f9 100644
--- a/Marlin/stepper.cpp
+++ b/Marlin/stepper.cpp
@@ -374,7 +374,7 @@ inline void update_endstops() {
           UPDATE_ENDSTOP(Y, MAX);
         #endif
       }
-  #if defined(COREXY) || defined(COREXZ)
+  #if defined(COREXY)
     }
   #endif
 
@@ -459,6 +459,9 @@ inline void update_endstops() {
           }
         #endif
       }
+  #if defined(COREXZ)
+    }
+  #endif  
   old_endstop_bits = current_endstop_bits;
 }
 

