commit a0a5ff61c4af0b8eb37f52ebbf33f33dab263431
Merge: 0094481 b59e92a
Author: Scott Lahteine <thinkyhead@users.noreply.github.com>
Date:   Thu Apr 23 18:40:00 2015 -0700

    Merge pull request #1951 from AnHardt/Azteeg_x3
    
    Fix for #1946

diff --git a/Marlin/pins_AZTEEG_X3.h b/Marlin/pins_AZTEEG_X3.h
index ea98f95..7c5dca3 100644
--- a/Marlin/pins_AZTEEG_X3.h
+++ b/Marlin/pins_AZTEEG_X3.h
@@ -28,9 +28,7 @@
    #define STAT_LED_RED 64
    #define STAT_LED_BLUE 63
   #endif
-#endif
-
-#elif define TEMP_STAT_LEDS
+#elif defined(TEMP_STAT_LEDS)
   #define STAT_LED_RED       6
   #define STAT_LED_BLUE     11
 #endif
