commit a1e8714d5fc8c70a76bbf61da5f2678e6b0c5f82
Merge: 149c474 b8b5762
Author: Richard Wackerbarth <rkw@dataplex.net>
Date:   Thu Jun 25 09:13:50 2015 -0500

    Merge 'Ultimaker' (PR#2331)

diff --git a/Marlin/language.h b/Marlin/language.h
index c852693..eb20a5b 100644
--- a/Marlin/language.h
+++ b/Marlin/language.h
@@ -46,7 +46,7 @@
 
 #if MB(ULTIMAKER)|| MB(ULTIMAKER_OLD)|| MB(ULTIMAIN_2)
   #define MACHINE_NAME "Ultimaker"
-  #define SOURCE_CODE_URL "http://firmware.ultimaker.com"
+  #define SOURCE_CODE_URL "https://github.com/Ultimaker/Marlin"
 #elif MB(RUMBA)
   #define MACHINE_NAME "Rumba"
 #elif MB(3DRAG)
diff --git a/Marlin/language_nl.h b/Marlin/language_nl.h
index 4e1ea6d..8211ca8 100644
--- a/Marlin/language_nl.h
+++ b/Marlin/language_nl.h
@@ -101,7 +101,7 @@
 #define MSG_DWELL                           "Slapen..."
 #define MSG_USERWAIT                        "Wachten..."
 #define MSG_RESUMING                        "Print hervatten"
-#define MSG_PRINT_ABORTED                   "Print aborted"
+#define MSG_PRINT_ABORTED                   "Print afgebroken"
 #define MSG_NO_MOVE                         "Geen beweging."
 #define MSG_KILLED                          "AFGEBROKEN. "
 #define MSG_STOPPED                         "GESTOPT. "
@@ -127,11 +127,11 @@
 #define MSG_END_MINUTE                      "minutes"
 
 #ifdef DELTA_CALIBRATION_MENU
-    #define MSG_DELTA_CALIBRATE             "Delta Calibration"
-    #define MSG_DELTA_CALIBRATE_X           "Calibrate X"
-    #define MSG_DELTA_CALIBRATE_Y           "Calibrate Y"
-    #define MSG_DELTA_CALIBRATE_Z           "Calibrate Z"
-    #define MSG_DELTA_CALIBRATE_CENTER      "Calibrate Center"
+    #define MSG_DELTA_CALIBRATE             "Delta Calibratie"
+    #define MSG_DELTA_CALIBRATE_X           "Kalibreer X"
+    #define MSG_DELTA_CALIBRATE_Y           "Kalibreer Y"
+    #define MSG_DELTA_CALIBRATE_Z           "Kalibreer Z"
+    #define MSG_DELTA_CALIBRATE_CENTER      "Kalibreer Midden"
 #endif // DELTA_CALIBRATION_MENU
 
 #endif // LANGUAGE_NL_H
