commit b099fe142cc6a62a2fccec78b4ac541612ebdf64
Merge: 8e017b8 67967a4
Author: Erik van der Zalm <erik@vdzalm.eu>
Date:   Wed Sep 14 18:57:32 2011 +0200

    Merge branch 'master' of github.com:ErikZalm/Marlin

diff --git a/README b/README
index daf03db..c11f90c 100644
--- a/README
+++ b/README
@@ -51,6 +51,20 @@ If all goes well the firmware is uploading
 Start pronterface
 
 Select the correct Serial Port. Type 250000 in the baudrate field.
-Press the Connect button.
+Press the Connect button
+
+===============================================================================================
+Known issues
+
+On some systems we get compilation errors.
+
+This is caused by the "wiring_serial.c" and "wiring.c".
+The simple fix is to delete these files but this may have a performance impact.
+
+The best workaround is to move these files to sanguino directory. 
+(".../arduino-0018/hardware/Sanguino/cores/arduino/" on windows systems)
+
+
+
 
 
