commit b84db7ff5b71f593b85f2b94e90fbc5d203f5416
Merge: fb74995 2966ae2
Author: Chris Roadfeldt <chris@roadfeldt.com>
Date:   Wed Apr 1 13:36:02 2015 -0500

    Merge branch 'Merge_cleanup' into Development

diff --git a/Marlin/pins.h b/Marlin/pins.h
index 6db56e9..0d63298 100644
--- a/Marlin/pins.h
+++ b/Marlin/pins.h
@@ -187,7 +187,7 @@
   #define Z_MIN_PIN          -1
 #endif
 
-#ifdef DISABLE_Z_PROBE_ENDSTOP
+#if defined (DISABLE_Z_PROBE_ENDSTOP) || ! defined (Z_PROBE_ENDSTOP) // Allow code to compile regardless of Z_PROBE_ENDSTOP setting.
   #define Z_PROBE_PIN        -1
 #endif
 
