commit c2f695c84e92ce53a941f183dd12f754760d1022
Merge: 9531b15 a1eb90d
Author: Richard Wackerbarth <rkw@dataplex.net>
Date:   Wed Aug 12 21:27:40 2015 -0500

    Merge Save a few bytes (PR#79)

diff --git a/Marlin/Marlin_main.cpp b/Marlin/Marlin_main.cpp
index 7c58ff7..fac6794 100644
--- a/Marlin/Marlin_main.cpp
+++ b/Marlin/Marlin_main.cpp
@@ -4786,31 +4786,29 @@ inline void gcode_M503() {
 #ifdef CUSTOM_M_CODE_SET_Z_PROBE_OFFSET
 
   inline void gcode_SET_Z_PROBE_OFFSET() {
-    float value;
+
+    SERIAL_ECHO_START;
+    SERIAL_ECHOPGM(MSG_ZPROBE_ZOFFSET);
+    SERIAL_CHAR(' ');
+
     if (code_seen('Z')) {
-      value = code_value();
+      float value = code_value();
       if (Z_PROBE_OFFSET_RANGE_MIN <= value && value <= Z_PROBE_OFFSET_RANGE_MAX) {
         zprobe_zoffset = value;
-        SERIAL_ECHO_START;
-        SERIAL_ECHOLNPGM(MSG_ZPROBE_ZOFFSET " " MSG_OK);
-        SERIAL_EOL;
+        SERIAL_ECHOPGM(MSG_OK);
       }
       else {
-        SERIAL_ECHO_START;
-        SERIAL_ECHOPGM(MSG_ZPROBE_ZOFFSET);
         SERIAL_ECHOPGM(MSG_Z_MIN);
         SERIAL_ECHO(Z_PROBE_OFFSET_RANGE_MIN);
         SERIAL_ECHOPGM(MSG_Z_MAX);
         SERIAL_ECHO(Z_PROBE_OFFSET_RANGE_MAX);
-        SERIAL_EOL;
       }
     }
     else {
-      SERIAL_ECHO_START;
-      SERIAL_ECHOPGM(MSG_ZPROBE_ZOFFSET " : ");
-      SERIAL_ECHO(zprobe_zoffset);
-      SERIAL_EOL;
+      SERIAL_ECHOPAIR(": ", zprobe_zoffset);
     }
+
+    SERIAL_EOL;
   }
 
 #endif // CUSTOM_M_CODE_SET_Z_PROBE_OFFSET
