commit d0d12962e0fa26f31d009f9ef3bc076161dbb3e9
Merge: 3f34d9d f0dcea7
Author: Scott Lahteine <thinkyhead@users.noreply.github.com>
Date:   Sun Mar 15 15:19:02 2015 -0700

    Merge pull request #1615 from AnHardt/fix-1611
    
    Fix #1611

diff --git a/Marlin/Marlin.h b/Marlin/Marlin.h
index 286a8d5..c056dd6 100644
--- a/Marlin/Marlin.h
+++ b/Marlin/Marlin.h
@@ -33,7 +33,7 @@
 #endif
 
 #define BIT(b) (1<<(b))
-#define TEST(n,b) ((n)&BIT(b)!=0)
+#define TEST(n,b) (((n)&BIT(b))!=0)
 
 // Arduino < 1.0.0 does not define this, so we need to do it ourselves
 #ifndef analogInputToDigitalPin
