commit fb74995fe3f1c48b2c4b2c15c831027276d17d80 (from f53e951462515df5dd36fca939a689c91809a233)
Merge: f53e951 bb41edc
Author: Chris Roadfeldt <chris@roadfeldt.com>
Date:   Wed Apr 1 12:15:01 2015 -0500

    Merge branch 'Merge_cleanup' into Development

diff --git a/Marlin/SanityCheck.h b/Marlin/SanityCheck.h
index cac7c55..6c0c9fd 100644
--- a/Marlin/SanityCheck.h
+++ b/Marlin/SanityCheck.h
@@ -124,7 +124,7 @@
 //      #endif
 //      #ifndef SERVO_ENDSTOP_ANGLES
 //        #error You must have SERVO_ENDSTOP_ANGLES defined for Z Extend and Retract to use Z_PROBE_AND_ENSTOP
-      #endif
+//      #endif
     #endif
     /**
      * Check if Probe_Offset * Grid Points is greater than Probing Range

